guvnor-ala/guvnor-ala-docker-provider/src/main/java/org/guvnor/ala/docker/executor/DockerRuntimeExecExecutor.java

```diff
@@ -0,0 +1,148 @@
+
+package org.guvnor.ala.docker.executor;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import com.spotify.docker.client.DockerException;
+import com.spotify.docker.client.messages.ContainerConfig;
+import com.spotify.docker.client.messages.ContainerCreation;
+import com.spotify.docker.client.messages.HostConfig;
+import com.spotify.docker.client.messages.PortBinding;
+import org.guvnor.ala.config.Config;
+import org.guvnor.ala.config.RuntimeConfig;
+import org.guvnor.ala.docker.access.DockerAccessInterface;
+import org.guvnor.ala.docker.config.DockerRuntimeExecConfig;
+import org.guvnor.ala.docker.model.DockerProvider;
+import org.guvnor.ala.docker.model.DockerRuntime;
+import org.guvnor.ala.exceptions.ProvisioningException;
+import org.guvnor.ala.pipeline.FunctionConfigExecutor;
+import org.guvnor.ala.registry.RuntimeRegistry;
+import org.guvnor.ala.runtime.RuntimeBuilder;
+import org.guvnor.ala.runtime.RuntimeDestroyer;
+import org.guvnor.ala.runtime.RuntimeId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.guvnor.ala.docker.config.DockerRuntimeConfig;
+
+public class DockerRuntimeExecExecutor<T extends DockerRuntimeConfig> implements RuntimeBuilder<T, DockerRuntime>,
+        RuntimeDestroyer,
+        FunctionConfigExecutor<T, DockerRuntime> {
+
+    private final RuntimeRegistry runtimeRegistry;
+    private final DockerAccessInterface docker;
+    protected static final Logger LOG = LoggerFactory.getLogger( DockerRuntimeExecExecutor.class );
+
+    @Inject
+    public DockerRuntimeExecExecutor( final RuntimeRegistry runtimeRegistry,
+            final DockerAccessInterface docker ) {
+        this.runtimeRegistry = runtimeRegistry;
+        this.docker = docker;
+    }
+
+    @Override
+    public Optional<DockerRuntime> apply( final DockerRuntimeConfig config ) {
+        final Optional<DockerRuntime> runtime = create( config );
+        if ( runtime.isPresent() ) {
+            runtimeRegistry.registerRuntime( runtime.get() );
+        }
+        return runtime;
+    }
+
+    private Optional<DockerRuntime> create( final DockerRuntimeConfig runtimeConfig ) throws ProvisioningException {
+        if ( runtimeConfig.isPull() ) {
+            try {
+                LOG.info( "Pulling Docker Image: " + runtimeConfig.getImage() );
+                docker.getDockerClient( runtimeConfig.getProviderId() ).pull( runtimeConfig.getImage() );
+            } catch ( DockerException | InterruptedException ex ) {
+                LOG.error( ex.getMessage(), ex );
+                throw new ProvisioningException( "Error Pulling Docker Image: " + runtimeConfig.getImage() + "with error: " + ex.getMessage() );
```

Comment:
I see that you log the caught exception, but maybe still add it as the cause for this thrown exception.